# **Exploratory Data Analysis on California Housing Dataset**



Loading and Exploring dataset properties



In [144]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('housing.csv')
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


**Datatypes of features**
* Categrorical : ocean_proximity
* Numerical : housing_median_age, total_rooms, total_bedrooms, population, households, median_income, longitude, latitude

In [146]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

One-Hot Encoding for 'ocean_proximity' and ensuring the new columns are of correct data type

In [147]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

In [148]:
df = pd.get_dummies(df, columns=['ocean_proximity'], drop_first=True)
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  \
0       322.0       126.0         8.3252            452600.0   
1      2401.0      1138.0         8.3014            358500.0   
2       496.0       177.0         7.2574            352100.0   
3       558.0       219.0         5.6431            341300.0   
4       565.0       259.0         3.8462            342200.0   

   ocean_proximity_INLAND  ocean_proximity_ISLAND  ocean_proximity_NEAR BAY  \
0                   False                   False                      True   
1                   False                   False                      True   
2                   False                   False                      True   
3                   False                   False                      True   
4                   False                   False                      True   

   ocean_proximity_NEAR OCEAN  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False

In [149]:
df[['ocean_proximity_INLAND', 'ocean_proximity_ISLAND', 'ocean_proximity_NEAR BAY', 'ocean_proximity_NEAR OCEAN']] = \
    df[['ocean_proximity_INLAND', 'ocean_proximity_ISLAND', 'ocean_proximity_NEAR BAY', 'ocean_proximity_NEAR OCEAN']].astype(float)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   longitude                   20640 non-null  float64
 1   latitude                    20640 non-null  float64
 2   housing_median_age          20640 non-null  float64
 3   total_rooms                 20640 non-null  float64
 4   total_bedrooms              20433 non-null  float64
 5   population                  20640 non-null  float64
 6   households                  20640 non-null  float64
 7   median_income               20640 non-null  float64
 8   median_house_value          20640 non-null  float64
 9   ocean_proximity_INLAND      20640 non-null  float64
 10  ocean_proximity_ISLAND      20640 non-null  float64
 11  ocean_proximity_NEAR BAY    20640 non-null  float64
 12  ocean_proximity_NEAR OCEAN  20640 non-null  float64
dtypes: float64(13)
memory usage: 2.

Filling missing values of 'total bedrooms' with median

In [150]:
df['total_bedrooms'].fillna(df['total_bedrooms'].median())
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   longitude                   20640 non-null  float64
 1   latitude                    20640 non-null  float64
 2   housing_median_age          20640 non-null  float64
 3   total_rooms                 20640 non-null  float64
 4   total_bedrooms              20433 non-null  float64
 5   population                  20640 non-null  float64
 6   households                  20640 non-null  float64
 7   median_income               20640 non-null  float64
 8   median_house_value          20640 non-null  float64
 9   ocean_proximity_INLAND      20640 non-null  float64
 10  ocean_proximity_ISLAND      20640 non-null  float64
 11  ocean_proximity_NEAR BAY    20640 non-null  float64
 12  ocean_proximity_NEAR OCEAN  20640 non-null  float64
dtypes: float64(13)
memory usage: 2.

Low correlation of 'total_bedrooms' and 'total_rooms' with target

In [151]:
df['total_bedrooms'].corr(df['median_house_value'])

np.float64(0.04968618024734591)

In [152]:
df['total_rooms'].corr(df['median_house_value'])

np.float64(0.1341531138065631)

High correlation between the two columns, risk of multicollinearity

In [153]:
df[['total_rooms', 'total_bedrooms']].corr()

total_rooms  total_bedrooms
total_rooms         1.00000         0.93038
total_bedrooms      0.93038         1.00000

New Feature: 'room_bedroom_ratio', has corr = 0.38

In [154]:
df['room_bedroom_ratio'] = df['total_rooms'] / df['total_bedrooms']
df['room_bedroom_ratio'].corr(df['median_house_value'])

np.float64(0.3839203207593848)

Low correlation of 'households' with target, introducing feature 'avg_rooms_per_household'

In [155]:
df['households'].corr(df['median_house_value'])

np.float64(0.06584265057005648)

In [156]:
df['avg_rooms_per_household'] = df['total_rooms'] / df['households']
df['avg_rooms_per_household'].corr(df['median_house_value'])

np.float64(0.15194828974145783)

Low correlation of 'population' with target, introducing feature 'people_per_household'

In [157]:
df['population'].corr(df['median_house_value'])

np.float64(-0.02464967888889488)

In [158]:
df['people_per_household'] = df['population'] / df['households']
df['people_per_household'].corr(df['median_house_value'])

np.float64(-0.023737412956134345)

In [159]:
df['people_per_household'].describe()

count    20640.000000
mean         3.070655
std         10.386050
min          0.692308
25%          2.429741
50%          2.818116
75%          3.282261
max       1243.333333
Name: people_per_household, dtype: float64

Low coorelation of 'people_per_household' without cap, and unsual max value at 1243, so we cap people_per_household at 99th percentile

In [160]:
cap_value = df['people_per_household'].quantile(0.99)
df['people_per_household'] = df['people_per_household'].clip(upper=cap_value)
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  ocean_proximity_INLAND  ocean_proximity_ISLAND  \
count        20640.000000            20640.000000            20640.000000   
mean        206855.816909                0.317393                0.000242   
std         115395.615874                0.465473                0.015563   
min          14999.000000                0.000000                0.000000   
25%         119600.000000                0.000000                0.000000   
50%         179700.000000                0.000000                0.000000   
75%         264725.000000                1.000000                0.000000   
max         500001.000000                1.000000                1.000000   

       ocean_proximity_NEAR BAY  ocean_proximity_NEAR OCEAN  \
count              20640.000000                20640.000000   
mean                   0.110950                    0.128779   
std                    0.314077                    0.334963   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                    0.000000                    0.000000   
75%                    0.000000                    0.000000   
max                    1.000000                    1.000000   

       room_bedroom_ratio  avg_rooms_per_household  people_per_household  
count        20433.000000             20640.000000          20640.000000  
mean             4.984829                 5.429000              2.915167  
std              1.171676                 2.474173              0.734751  
min              1.000000                 0.846154              0.692308  
25%              4.169782                 4.440716              2.429741  
50%              4.922170                 5.229129              2.818116  
75%              5.700364                 6.052381              3.282261  
max             10.000000               141.909091              5.394812

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   longitude                   20640 non-null  float64
 1   latitude                    20640 non-null  float64
 2   housing_median_age          20640 non-null  float64
 3   total_rooms                 20640 non-null  float64
 4   total_bedrooms              20433 non-null  float64
 5   population                  20640 non-null  float64
 6   households                  20640 non-null  float64
 7   median_income               20640 non-null  float64
 8   median_house_value          20640 non-null  float64
 9   ocean_proximity_INLAND      20640 non-null  float64
 10  ocean_proximity_ISLAND      20640 non-null  float64
 11  ocean_proximity_NEAR BAY    20640 non-null  float64
 12  ocean_proximity_NEAR OCEAN  20640 non-null  float64
 13  room_bedroom_ratio          204

We now drop the unnecessary columns

In [165]:
columns_to_drop = ['total_rooms', 'total_bedrooms', 'population', 'households']
df.drop(columns=columns_to_drop, inplace=True)
df.info()

KeyError: "['total_rooms', 'total_bedrooms', 'population', 'households'] not found in axis"

The Preprocessed data is now exported to a csv file 'data.csv'

In [164]:
target = df['median_house_value']
df = df.drop(columns=['median_house_value'])
df['median_house_value'] = target
df.to_csv('data.csv', index=False)

KeyError: 'median_house_value'